doc/design/determinism_annotation.md

```diff
@@ -0,0 +1,58 @@
+# Determinism Annotations
+
+## Cert
+
+All functions are to be certified, hence there will be no label to indicate this.
+
+## Deterministic
+
+`\deterministic` indicates that the function is real-time safe, i.e. returns in a finite amount of
+time which is (somewhat) predictable.
+
+In particular this rules out.
+
+1. not using exceptions
+2. no dynamic allocations from an external allocator (new, malloc)
+3. non-blocking
+
+(TBD: Non blocking may still use blocking calls, there must be a return guarantee though.)
+
+Note: deterministic is the wrong wording and should be changed.
+
+## Nondeterminstic
+
+Will not be labeled on its own. Instead we have labels indicating the details.
+
+### Blocking
+
+`\blocking` indicates the code will block and needs o be unblocked externally. Say a mutex lock that
+needs to be unlocked from outside the call (it it is unlocked in the call it is not blocking).
+
+A loop that waits for a conition to become true is also blocking if the condition cannot become true
+due to actions of the function itself.
+
+### Dynamic Memory
+`\allocating` indicates it will allocate memory and implies `\dynamic`
+`\dynamic` indicates the code will invoke dynamic memory functions.
+
+Invoking free does count as well. This can be debated, free cannot fail but it will cause an
+invocation of the allocator which is unpredictable.
+
+Allocating may also fail.
+
+To clarify: use for all resources (like e.g. semaphores?)
+
+### Throwing
```

Comment:
Is this really necessary? There is a nice way of showing this with plain C++ syntax called `noexcept`. From C++17 and onwards this is even part of the function signature.

Reply:
@mossmaurice The intention is different but the term in not well chosen. Besides, noexcept does not tell us this. It can be throwing but noexcept. Then it terminates.

For real-time/determinism purposes it is not only about whether an exception can leave the function (it cannot, with noexcept), but whether the exception mechanism is used at all (since it introduces allocation and unpredictable overhead). noexcept does not capture this.

Two situations that are possible with throwing
1) exception is thrown in the function but also caught there.
2) exception is thrown and leaves the function (and terminates the program, if noexcept).

For a non-throwing function neither is allowed, no exceptions are used in any way. The deterministic tag requires the function to be non-throwing  (the name deterministic is bad here but there is some reason for it ...).

But what if it is not deterministic? Then there is a reason, and throwing exceptions is one possible reason. It will not really hold for any of our own code though (careful with custom types and callbacks, they must be restricted).
But the idea is, if it is not